app/src/main/java/org/odk/odknotifications/DatabaseCommunicator/ServerDatabaseCommunicator.java

```diff
@@ -0,0 +1,261 @@
+
+        String roles_array_string = userDbInterface.getRolesList(appName);
+
+        JSONArray temp = new JSONArray();
+
+        if(roles_array_string!=null) {
+            temp = new JSONArray(roles_array_string);
+        }
+
+        StringBuilder groups = new StringBuilder();
+        for(int i=0; i<temp.length();i++){
+            if(temp.getString(i).startsWith("GROUP_")|| temp.getString(i).startsWith("ROLE_")) {
+                groups.append(temp.get(i)).append(",");
+            }
+        }
+
+        String userName = getUserName(userId);
+
+        List<String>columnValues = Arrays.asList(userId,userName,groups.toString());
+
+        ContentValues contentValues = new ContentValues();
+
+        for(int i=0;i<columnValues.size();i++){
+            contentValues.put(USERS_TABLE_COLUMNS_LIST.get(i) , columnValues.get(i));
+        }
+
+        OrderedColumns orderedColumns = userDbInterface.getUserDefinedColumns(appName,dbHandle,USER_TABLE_ID);
+
+        userDbInterface.insertRowWithId(appName,dbHandle,USER_TABLE_ID,orderedColumns,contentValues,userId);
+    }
+
+    private static Group getGroupFromId(String groupId) throws ServicesAvailabilityException {
+        Group group = new Group();
+
+        OrderedColumns orderedColumns = userDbInterface.getUserDefinedColumns(appName,dbHandle,GROUPS_TABLE_ID);
+
+        UserTable userTable = userDbInterface.simpleQuery(appName, dbHandle, GROUPS_TABLE_ID, orderedColumns, null, null,
+                null,null,null,null,null,null);
+
+        int rowNo = userTable.getRowNumFromId(groupId);
+
+        if(rowNo >= 0) {
+            TypedRow typedRow = userTable.getRowAtIndex(rowNo);
+            group.setId(typedRow.getStringValueByKey("GroupId"));
+            group.setName(typedRow.getStringValueByKey("GroupName"));
+        }
+        return group;
+    }
+
+    private static String getResponse(String notificationId) throws ServicesAvailabilityException {
+
+        String response = "";
+
+        OrderedColumns orderedColumns = userDbInterface.getUserDefinedColumns(appName,dbHandle,RESPONSES_TABLE_ID);
+
+        UserTable userTable = userDbInterface.simpleQuery(appName, dbHandle, RESPONSES_TABLE_ID, orderedColumns, null, null,
+                null,null,null,null,null,null);
+
+        for(int i=0;i<userTable.getNumberOfRows();i++){
+            TypedRow typedRow = userTable.getRowAtIndex(i);
+
+            if(typedRow.getStringValueByKey("NotificationId").equals(notificationId) &&
+                    typedRow.getStringValueByKey("UserName").equals(getUserName(userId))){
+                response = typedRow.getStringValueByKey("ResponseText");
+            }
+        }
+
+        return response;
+    }
+
+    private static String getUserName (String userId){
+        String userName = userId;
+
+        if(!(userId.compareTo("anonymous")==0) && userId.length()>8 && userId.substring(0,9).compareTo("username:")==0){
```

Comment:
Same as commented above. 
And try using some constants of you're doing same operation at different places on same string.
And here it seems both this and one of the methods above I commented in also does the same thing, so its better to have a single method and then both of these places will use that method

Reply:
Added constant strings and comment for using substring operation. 
This method and one you commented above are from two different classes. So do you want me to make this method public and use it in another class?